### PR TITLE
receiver/prometheus: add store to track stale metrics

### PR DESCRIPTION
**Pull request by odeke-em**
Implements a simple store that will be used to record stale metrics;
those that are NOT present in the current scrape, but were in the prior
scrape. This logic will then be attached to our receiver logic
to then complete the mechanism and then issue stale markers.

Updates #3413
Updates #2961

### CODE REVIEW

**Comment by odeke-em** [comment]
If you'd like the big picture, please see https://github.com/orijtech/opentelemetry-collector/pull/new/receiver-prometheus-add-staleness-tracking+emit-staleness-markers and after it the remotewrite compliance test for Staleness passes with
```shell
        --- PASS: TestRemoteWrite/otelcollector/Staleness (10.04s)
```

Kindly cc-ing related reviewers @dashpole @rakyll @Aneurysm9 @bogdandrutu @tigrannajaryan.

**Review by Aneurysm9 — APPROVED** [review]

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (dashpole) [inline receiver/prometheusreceiver/internal/staleness_store.go:38]: If you want to ensure performance doesn't regress, it might be useful to add a benchmark.

**Review by dashpole — APPROVED** [review]

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (odeke-em) [reply receiver/prometheusreceiver/internal/staleness_store.go:38]: It's more of Go compiler magic to let future code maintainers to know the importance of it as few people know about this. Adding benchmarks won't add much value honestly, the comment does suffice IMHO.

**Comment by odeke-em** [comment]
@bogdandrutu @tigrannajaryan the failed check seems to be  flake from the CI itself and unrelated to code:
```shell
6d54c59f90a1c17832fdec2373b613351ac3046a5aa62f66c0e127895e61e5f7
Installing otel-collector_0.28.0-post_amd64.deb ...
Error response from daemon: Error processing tar file(exit status 1): lchown /tmp/otel-collector_0.28.0-post_amd64.deb: no such file or directory
Error: Process completed with exit code 1.
```

**Comment by odeke-em** [comment]
'The PR to finish things up is https://github.com/open-telemetry/opentelemetry-collector/pull/3423 but unfortunately Github doesn't stack/link changes so that PR has code from this one, but after this one is merged then it'll get to its normal size.

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (rakyll) [inline receiver/prometheusreceiver/internal/staleness_store.go:35]: It's not clear to me whose responsibility it is to call refresh. Can you add to the method doc that refresh is called after scrape. The tests are doing a better job how to use the stalenessStore but it's not clear from the package's itself.

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (rakyll) [inline receiver/prometheusreceiver/internal/staleness_store.go:62]: Maybe add a comment saying that this method needs to be called before refresh if a scrape has seen a label.

**Review by rakyll — APPROVED** [review]

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (odeke-em) [reply receiver/prometheusreceiver/internal/staleness_store.go:63]: Done, thanks!

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (odeke-em) [reply receiver/prometheusreceiver/internal/staleness_store.go:35]: Added, thank you, PTAL.

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (tigrannajaryan) [inline receiver/prometheusreceiver/internal/staleness_store.go:21]: Please add a comment to explain the purpose and the intended usage. (Yes, I know the functions below have descriptions. I am looking for something that is more of an overall description of how the store should be used and why).
Links to previous discussions or issues which describe the problem can be also useful.

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (rakyll) [reply receiver/prometheusreceiver/internal/staleness_store.go:39]: Is there a reference you can give to explain the optimization? Not every maintainer is knowledgeable about this and would be useful to have a reference.

Comment (odeke-em) [reply receiver/prometheusreceiver/internal/staleness_store.go:43]: Done, please take a look.

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (odeke-em) [reply receiver/prometheusreceiver/internal/staleness_store.go:25]: Done, thanks @tigrannajaryan!

**Comment by alolita** [comment]
@odeke-em some of the required tests are failing. Pl take a look and fix.

**Comment by odeke-em** [comment]
@alolita, the Windows test failure is a flake from an entirely different package that we haven't touched per https://github.com/open-telemetry/opentelemetry-collector/pull/3414/checks?check_run_id=2823559150#step:6:179
```shell
ok  	go.opentelemetry.io/collector/config/configauth	0.029s
# go.opentelemetry.io/collector/config/configgrpc [go.opentelemetry.io/collector/config/configgrpc.test]
config\configgrpc\configgrpc_test.go:456:38: cannot use &grpcTraceServer{} (type *grpcTraceServer) as type pdatagrpc.TracesServer in argument to pdatagrpc.RegisterTracesServer:
	*grpcTraceServer does not implement pdatagrpc.TracesServer (wrong type for Export method)
		have Export(context.Context, pdata.Traces) (interface {}, error)
		want Export(context.Context, pdata.Traces) (pdatagrpc.TracesResponse, error)
config\configgrpc\configgrpc_test.go:501:36: cannot use &grpcTraceServer{} (type *grpcTraceServer) as type pdatagrpc.TracesServer in argument to pdatagrpc.RegisterTracesServer:
	*grpcTraceServer does not implement pdatagrpc.TracesServer (wrong type for Export method)
		have Export(context.Context, pdata.Traces) (interface {}, error)
		want Export(context.Context, pdata.Traces) (pdatagrpc.TracesResponse, error)
```

same for the lint failures being unrelated to this change
```shell
make[1]: Leaving directory '/home/runner/work/opentelemetry-collector/opentelemetry-collector'
config/configgrpc/configgrpc_test.go:456:38: cannot use &(grpcTraceServer literal) (value of type *grpcTraceServer) as pdatagrpc.TracesServer value in argument to pdatagrpc.RegisterTracesServer: wrong type for method Export (have func(context.Context, go.opentelemetry.io/collector/consumer/pdata.Traces) (interface{}, error), want func(context.Context, go.opentelemetry.io/collector/consumer/pdata.Traces) (go.opentelemetry.io/collector/internal/pdatagrpc.TracesResponse, error)) (typecheck)
			pdatagrpc.RegisterTracesServer(s, &grpcTraceServer{})
			                                  ^
config/configgrpc/configgrpc_test.go:501:36: cannot use &(grpcTraceServer literal) (value of type *grpcTraceServer) as pdatagrpc.TracesServer value in argument to pdatagrpc.RegisterTracesServer: wrong type for method Export (have func(context.Context, go.opentelemetry.io/collector/consumer/pdata.Traces) (interface{}, error), want func(context.Context, go.opentelemetry.io/collector/consumer/pdata.Traces) (go.opentelemetry.io/collector/internal/pdatagrpc.TracesResponse, error)) (typecheck)
	pdatagrpc.RegisterTracesServer(s, &grpcTraceServer{})
	                                  ^
make[2]: *** [Makefile.Common:26: lint] Error 1
```

**Comment by odeke-em** [comment]
All tests now passing on their own after the flakes.

File: receiver/prometheusreceiver/internal/staleness_store.go
Comment (tigrannajaryan) [reply receiver/prometheusreceiver/internal/staleness_store.go:25]: 👍

**Review by tigrannajaryan — APPROVED** [review]

**Pull request merged**
